### PR TITLE
Add a Script to generate changelog

### LINKED ISSUE #1039: Render MarkupText colors using <span> tags
Currently, coloring uses `<color>` tags which are Manim-specific (unsupported by Pango), and were a workaround for lack of color parsing in the SVG engine. Now that SVG parsing for fill and color has improved (#915), MarkupText can pass a foreground color to Pango, and Pango's "official" coloring tag, `<span foreground="#rrggbb">`, can be used.

As discussed later in #915, there are two parts:
1. Passing in the fill color (which fill color?) to Pango as the `AttrForeground` attribute.
2. Removing the `init_colors` override for MarkupText objects.

cc @PhilippImhof

### LINKED ISSUE #1045: Broken source links on the stable version of the documentation
## Description of bug / unexpected behavior
Source links on the stable version of documentation does not work. It links to something like this: https://github.com/ManimCommunity/manim/blob/stable/manim/mobject/changing.py which is a 404 error. 

## Expected behavior
Source links should link to a file containing source code for the stable version. 

## How to reproduce the issue
On the documentation website, switch the version to stable. Navigate to and click the source link of any class. 

## Additional comments
Perhaps this is an access rights issue, which explains why it evaded detection from community devs for so long?

### PR DESCRIPTION
**Pull request by naveen521kk**
Originally taken from Numpy
This will be used to generate changelogs.

## Motivation
Fixes https://github.com/ManimCommunity/manim/issues/1091
Tried towncrier but it wasn't useful enough that the Numpy(which we initially looked at) doesn't use it. 

This script quickly generates a changelog for whoever is making the release.

This will require some amount of manual work and isn't as simple as tagging a release on Github.
The script is well documented and I will update the instructions on the Wiki page later.

## Testing Status
Works locally
<details><summary>Logs</summary>

```sh
$ python scripts/dev_changelog.py $GITHUB_TOKEN v0.3.0...v0.4.0

Contributors
============

A total of 20 people contributed to this release.  People with a "+" by their
names contributed a patch for the first time.

* Aathish Sivasubrahmanian
* Abhijith Muthyala
* Anoop Hallur +
* Benjamin Hackl
* Clar Fon +
* Darylgolden +
* Devin Neal
* Jason Villanueva
* Kapil Sachdeva +
* KingWampy
* M. Eugenia Moreno +
* Mark Miller
* Naveen M K
* Sergii Penner +
* SuperMaZingCoder +
* friedkeenan
* kolibril13
* sparshg +
* yalikes +
* ▒lvaro Mond▒jar +

Pull requests merged
====================

A total of 41 pull requests were merged for this release.

* `#915 <https://github.com/ManimCommunity/manim/pull/915>`__: SVG engine rewrite and tests
* `#969 <https://github.com/ManimCommunity/manim/pull/969>`__: fix the problem that --tex_template flag didn't handle.
* `#981 <https://github.com/ManimCommunity/manim/pull/981>`__: Bugfix: Fix WebGL renderer hot reload on Windows
* `#983 <https://github.com/ManimCommunity/manim/pull/983>`__: Fix last frame issue
* `#984 <https://github.com/ManimCommunity/manim/pull/984>`__: Make the Manim version used more apparent
* `#989 <https://github.com/ManimCommunity/manim/pull/989>`__: Fix 'manim cfg export' subcommand
* `#993 <https://github.com/ManimCommunity/manim/pull/993>`__: can't pass list of colors in set_color() fixed
* `#995 <https://github.com/ManimCommunity/manim/pull/995>`__: Add generic set method and compatibility layer between properties...
* `#996 <https://github.com/ManimCommunity/manim/pull/996>`__: Edit message in --version for consistency
* `#998 <https://github.com/ManimCommunity/manim/pull/998>`__: Fix error using #hex colors with 3 characters
* `#1000 <https://github.com/ManimCommunity/manim/pull/1000>`__: Remove 'MovingCameraScene.setup' and 'MovingCameraScene.camera_frame'
* `#1003 <https://github.com/ManimCommunity/manim/pull/1003>`__: GrowArrow animation fixed
* `#1005 <https://github.com/ManimCommunity/manim/pull/1005>`__: Fix - as filename feature
* `#1008 <https://github.com/ManimCommunity/manim/pull/1008>`__: ci: fix release asset upload
* `#1010 <https://github.com/ManimCommunity/manim/pull/1010>`__: Disable STDIN interaction for ffmpeg concat.
* `#1015 <https://github.com/ManimCommunity/manim/pull/1015>`__: Add steps to reboot computer and update tlgmr.
* `#1017 <https://github.com/ManimCommunity/manim/pull/1017>`__: Added progress_bar to digest_args
* `#1018 <https://github.com/ManimCommunity/manim/pull/1018>`__: Remove redundancy in FunctionGraph arguments
* `#1021 <https://github.com/ManimCommunity/manim/pull/1021>`__: Docs: added Bezier example
* `#1022 <https://github.com/ManimCommunity/manim/pull/1022>`__: Fix using -p and -s together
* `#1024 <https://github.com/ManimCommunity/manim/pull/1024>`__: Migrate width/height/depth to properties
* `#1026 <https://github.com/ManimCommunity/manim/pull/1026>`__: Add 3D Mobjects - Cone, Cylinder, Line3D, Arrow3D and Torus.
* `#1028 <https://github.com/ManimCommunity/manim/pull/1028>`__: Update mac installation on Mac with apple silicon
* `#1030 <https://github.com/ManimCommunity/manim/pull/1030>`__: Enhancement: parse .log file and try to print LaTeX errors if...
* `#1031 <https://github.com/ManimCommunity/manim/pull/1031>`__: Fix typo to link in doc
* `#1032 <https://github.com/ManimCommunity/manim/pull/1032>`__: Remove Square.side_length property
* `#1043 <https://github.com/ManimCommunity/manim/pull/1043>`__: Breaking: Make CubicBezier explicitly accept four points
* `#1044 <https://github.com/ManimCommunity/manim/pull/1044>`__: register_font is available for macOS
* `#1046 <https://github.com/ManimCommunity/manim/pull/1046>`__: Update importlib-metadata
* `#1047 <https://github.com/ManimCommunity/manim/pull/1047>`__: add api docs and examples for Matrix mobjects
* `#1048 <https://github.com/ManimCommunity/manim/pull/1048>`__: fix - use absolute path in make_and_open_docs.py
* `#1050 <https://github.com/ManimCommunity/manim/pull/1050>`__: fix - make sure stroke color & width gets applied in Cross object
* `#1051 <https://github.com/ManimCommunity/manim/pull/1051>`__: fix - corrections for setting stroke related attributes on VMobject
* `#1053 <https://github.com/ManimCommunity/manim/pull/1053>`__: fix #1045 changes in line 118
* `#1058 <https://github.com/ManimCommunity/manim/pull/1058>`__: Fix #1039 (<span foreground=> tags)
* `#1059 <https://github.com/ManimCommunity/manim/pull/1059>`__: More descriptive error when accessing an unhandled mobject attribute
* `#1060 <https://github.com/ManimCommunity/manim/pull/1060>`__: docs: update installation docs for linux - pango
* `#1063 <https://github.com/ManimCommunity/manim/pull/1063>`__: Fix docs related to .animate
* `#1065 <https://github.com/ManimCommunity/manim/pull/1065>`__: Remove duplicate word 'vector'
* `#1067 <https://github.com/ManimCommunity/manim/pull/1067>`__: rtd: add manimpango wheels
* `#1072 <https://github.com/ManimCommunity/manim/pull/1072>`__: Prepare v0.4.0
```
</details>

## Acknowledgements
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have chosen a descriptive PR title (see top of PR template for examples)

## Reviewer Checklist
- [ ] Newly added functions/classes are either private or have a docstring
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings
- [ ] The PR title is descriptive enough

### CODE REVIEW

**Comment by WampyCakes** [comment]
With the addition of this I think we ought to also make a short standardized way of naming PRs. If every PR starts with an identifier like "Bugfix:" or "New Feature:" or "Refactor:" etc. I think it would help us to be more organized (this could either be directly written in the PR template or have a link to our wiki or docs with the information for naming). To show a bad example, my own PR was poorly named because of the redundancy between "Bugfix" and "Fix." "Bugfix: Fix WebGL renderer hot reload on Windows"

Another thing I thought of recently is that imo one of the changelog headers should be documentation. I think it makes more sense for new or changing documentation to be listed under its own section.

**Comment by jsonvillanueva** [comment]
I agree with @WampyCakes, especially because we have opted to NOT use towncrier which would force the PR author to create their own "fragments" with the appropriate extension. I'm wondering if we can make this script better by retrieving the label information from the PR so that we can still automatically organize the PRs into sections without changing the workflow of the PR author.

**Review by jsonvillanueva — APPROVED** [review]

**Comment by WampyCakes** [comment]
I like the idea of using labels. Before merging, an assortment of labels that could be chosen by the merger could go a long way in putting PRs under the correct headers. Obviously this would not be a sufficient replacement to well-named PRs though.

**Comment by GameDungeon** [comment]
I wonder if we could add a GitHub test to make sure the title has a tag as we do with black. We would need a standard, but we should open a new issue for that.

**Comment by naveen521kk** [comment]
I think what we can do is add a section in the PR template which the script picks as a changelog seems to be a better idea. thoughts?

**Comment by jsonvillanueva** [comment]
I think a better method is to use GitHub's API. For instance, for this PR:
https://api.github.com/repos/ManimCommunity/manim/issues/1138
You can navigate through the json to get the appropriate label names and sort by precedence:
```
  "labels": [
    {
      "id": 2306291174,
      "node_id": "MDU6TGFiZWwyMzA2MjkxMTc0",
      "url": "https://api.github.com/repos/ManimCommunity/manim/labels/easy%20review",
      "name": "easy review",
      "color": "bfd4f2",
      "default": false,
      "description": "There is nothing particular (i.e, it's about a general/small thing) to know for review!"
    },
    {
      "id": 2445988560,
      "node_id": "MDU6TGFiZWwyNDQ1OTg4NTYw",
      "url": "https://api.github.com/repos/ManimCommunity/manim/labels/release",
      "name": "release",
      "color": "000000",
      "default": false,
      "description": "A tracking issue for changes expected for a release"
    }
  ],
  ```
In this manner, the sections can be organized by the appropriate flags. If the PR has multiple though, we would  need to decide which section they go under/tiebreakers. We already have the PR #s from this script so this wouldn't be that difficult to implement given a python json parser. **But for a future PR.**

**Comment by naveen521kk** [comment]
Let the labels be like, everything with a common prefix say `change`
- highlight
- deprecation
- new_feature
- improvement
- breaking
- docs
- other-changes

A PR can be marked one or more than one, and also PR authors will write a description in PR body starting like
```md
## Changelog
<!--changelog-starts-->
```rst
[Changelog goes here]
```\
<!--changelog-ends-->
```
where we will look for the two comments and pick the code block and paste it in the changelog, under appropriate sections as labels say the section. Thoughts?

**Comment by jsonvillanueva** [comment]
> and also PR authors will write a description in PR body starting like

I don't like this because forcing the user to write an additional `rst` description (beyond the title):
1. is additional work on the PR author
2. involves some sort of pattern matching for the PR that might not exist because the PR author didn't write one
3. forces them to learn to write in `rst` format.
4. seems like re-implementing towncrier -- in which case, just use towncrier/have people write fragments.

Using GitHub's API seems best imo. A PR will *ALWAYS* have a title (it should be descriptive enough if the reviewers follow the checklist) and can easily have labels added. If the title/label is edited after the PR is merged, the changelog can still be automatically regenerated using the updated values from the API.

**Comment by naveen521kk** [comment]
> I don't like this because forcing the user to write an additional rst description (beyond the title):

I think you meant the devs. I would say it is fine to do so because they are the people who wrote the code and know what it does. Maybe Reviewers can help in writing them.

> involves some sort of pattern matching for the PR that might not exist because the PR author didn't write one

In that case, it is about just falling back to the PR title.

> forces the PR author to learn to write in rst format.

I would say it is much necessary, RST format is used everywhere in docs and also in docstrings. MD should be fine if necessary changes are made, if that's an issue.

> seems like re-implementing towncrier -- in which case, why not just use towncrier/have people write fragments.

I originally liked the idea of towncrier, but it doesn't work as expected sadly :(
see Numpy's changelog where they didn't use towncrier in the latest release because it doesn't work.

> Using GitHub's API seems best imo. A PR will ALWAYS have a title (it should be descriptive enough if the reviewers follow the checklist) and can easily have labels added. If the title/label is edited after the PR is merged, the changelog can still be automatically regenerated using the updated values from the API.

Doing this also needs to use the Github API. I am just adding extra info to your idea. Use the labels to organise into sections and have some info or detailed changelog more than just the title.

**Comment by jsonvillanueva** [comment]
> I think you meant the devs. I would say it is fine to do so because they are the people who wrote the code and know what it does. Maybe Reviewers can help in writing them.

Yes, I meant the PR author/dev. 
> In that case, it is about just falling back to the PR title.

Okay, I'm starting to like the idea more because of the fallback, but I still think it should be optional. 
> I would say it is much necessary, RST format is used everywhere in docs and also in docstrings. MD should be fine if necessary changes are made, if that's an issue.

While I agree it is very useful (if not necessary), GitHub Flavored Markdown would be better because PR authors won't have to manually verify their RST is formatted properly outside of GitHub. Until #1137 adds the documentation to the website, this doesn't much sense to do with RST because some contributors don't realize there is a GitHub Wiki page for our documentation -- instead they've used a plugin (#1064) or just copied the documentation of similar items already in the docs.

> I originally liked the idea of towncrier, but it doesn't work as expected sadly :(
> see Numpy's changelog where they didn't use towncrier in the latest release because it doesn't work.

I was wondering why they didn't when I was working on towncrier last week. I know towncrier hasn't had a latest release in a very long time so it has to be downloaded from latest/put into the pyproject.toml as a URL. I didn't look much further into towncrier, but I was still able to generate new fragments for different sections with `towncrier create <#>.<feature/bugfix/doc/etc.>`

> Doing this also needs to use the Github API. I am just adding extra info to your idea. Use the labels to organise into sections and have some info or detailed changelog more than just the title.

In general, this *can* work and the extra information from the detailed changelog/description is useful if the title isn't enough space, but it would need testing to make sure it works at release time since we just need it to be automatic. To make this easier, maybe it shouldn't be RST or GFD, but just a simple paragraph.

Eitherway, we *could* still do the label organization for this release. We still have ~1 week to label the merged PRs for this release. I think it'd be too much work to add in paragraph descriptions for the merged PRs, or test that it properly gets the changelog paragraph description properly/fallsback.

**Comment by naveen521kk** [comment]
> Eitherway, we could still do the label organization for this release. We still have ~1 week to label the merged PRs for this release. I think it'd be too much work to add in paragraph descriptions for the merged PRs, or test that it properly gets the changelog paragraph description properly/fallsback.

BTW, for the towncrier I already spend some amount of time https://github.com/naveen521kk/manim/commit/0789af302f9cef1a4bdf2e2be274495484957c43 in writing PR desc.

> While I agree it is very useful (if not necessary), GitHub Flavored Markdown would be better because PR authors won't have to manually verify their RST is formatted properly outside of GitHub. 

We can use some kind of project to convert those markdown to rst format.

> Until #1137 adds the documentation to the website, this doesn't much sense to do with RST because some contributors don't realize there is a GitHub Wiki page for our documentation -- instead they've used a plugin (#1064) or just copied the documentation of similar items already in the docs.

We have a link to the wiki in the PR template.

> I was wondering why they didn't when I was working on towncrier last week. I know towncrier hasn't had a latest release in a very long time so it has to be downloaded from latest/put into the pyproject.toml as a URL. I didn't look much further into towncrier, but I was still able to generate new fragments for different sections with towncrier create <#>.<feature/bugfix/doc/etc.>

The final generated changelog is broken you can try it out https://github.com/naveen521kk/manim/commit/0789af302f9cef1a4bdf2e2be274495484957c43

>

**Comment by jsonvillanueva** [comment]
On that note, it took ~15 minutes to go through the already merged PRs and add the appropriate labels to those missing them.

**Comment by jsonvillanueva** [comment]
> We have a link to the wiki in the PR template.

We do, but it's under the reviewer section -- I think people see the part that says "" and then just don't bother reading the rest. We should update the PR_Template.md to have the same check list for the author

**Comment by naveen521kk** [comment]
I will work on this tomorrow or the day after.

**Comment by jsonvillanueva** [comment]
I'm currently working on it, will push if I get it neatly formatted rst -- otherwise my changes will be in https://github.com/jsonvillanueva/manim/tree/changelog for you to work from

**Comment by jsonvillanueva** [comment]
<details>
<summary>Was able to generate this changelog after tagging locally a v0.5.0</summary>
```


Contributors
============

A total of 24 people contributed to this release.
People with a '+' by their names contributed a patch for the first time.

* Abel Aebker +
* Abhijith Muthyala
* AntonBallmaier +
* Aron
* Bogdan Stăncescu +
* Darylgolden
* Devin Neal
* GameDungeon +
* Hugues Devimeux
* Jason Villanueva
* KingWampy
* Lionel Ray +
* Mark Miller
* Mohammad Al-Fetyani +
* Naveen M K
* Niklas Dewally +
* Oliver +
* Roopesh +
* Seb Pearce +
* aebkea +
* friedkeenan
* hydrobeam +
* kolibril13
* tfglynn +

Pull requests merged
====================

A total of 34 pull requests were merged for this release.

Breaking changes
----------------

Highlight
---------

New feature
-----------

* `#1113 <https://github.com/ManimCommunity/manim/pull/1113>`__: Added interaction to the OpenGL renderer
* `#1074 <https://github.com/ManimCommunity/manim/pull/1074>`__: Added jupyter media_width option to the config
* `#1107 <https://github.com/ManimCommunity/manim/pull/1107>`__: Added :class:`~.Unwrite` animation class to complement :class:`~.Write`
* `#1085 <https://github.com/ManimCommunity/manim/pull/1085>`__: Added Angle/RightAngle classes for two intersecting lines
Enhancement
-----------

* `#1144 <https://github.com/ManimCommunity/manim/pull/1144>`__: Improved quality of GIFs
* `#1157 <https://github.com/ManimCommunity/manim/pull/1157>`__: Refresh triangulation on call to :meth:`~.OpenGLVMobject.apply_points_function`
* `#1151 <https://github.com/ManimCommunity/manim/pull/1151>`__: Added parametric function support to :class:`OpenGLSurface`
* `#1139 <https://github.com/ManimCommunity/manim/pull/1139>`__: In-Code `config["preview"]` Support
* `#1123 <https://github.com/ManimCommunity/manim/pull/1123>`__: Added caching, skipping, and user-specified background colors to the OpenGL renderer
* `#1118 <https://github.com/ManimCommunity/manim/pull/1118>`__: Allow passing animation arguments with .animate syntax
* `#718 <https://github.com/ManimCommunity/manim/pull/718>`__: Rotating the numbers in y axis
* `#1070 <https://github.com/ManimCommunity/manim/pull/1070>`__: Raise FileNotFoundError when unable to locate the .cfg file specified via `--config_file`
Bug
---

* `#1134 <https://github.com/ManimCommunity/manim/pull/1134>`__: Fixed the `-a` flag.
* `#1115 <https://github.com/ManimCommunity/manim/pull/1115>`__: Fixed bugs in :class:`~.OpenGLMobject` and added :class:`ApplyMethod` support 
* `#1092 <https://github.com/ManimCommunity/manim/pull/1092>`__: Refactored coordinate_systems.py, fixed bugs, added :class:`~.NumberPlane` test
* `#1078 <https://github.com/ManimCommunity/manim/pull/1078>`__: Removed stray print statements from `__main__.py`
* `#1086 <https://github.com/ManimCommunity/manim/pull/1086>`__: Fixed broken line spacing in :class:`~.Text`
* `#1083 <https://github.com/ManimCommunity/manim/pull/1083>`__: Corrected the shape of :class:`~.Torus`
Documentation
-------------

* `#1077 <https://github.com/ManimCommunity/manim/pull/1077>`__: Added details to :func:`~.Mobject.get_critical_point`
* `#1154 <https://github.com/ManimCommunity/manim/pull/1154>`__: Fixed some typing hints. (ints to floats)
* `#1036 <https://github.com/ManimCommunity/manim/pull/1036>`__: Added :class:`~.SurroundingRectangle` to the example gallery
* `#1103 <https://github.com/ManimCommunity/manim/pull/1103>`__: Added documentation and examples for Square, Dot, Circle and Rectangle
* `#1101 <https://github.com/ManimCommunity/manim/pull/1101>`__: Added documentation to :class:`~.Mobject`
* `#1088 <https://github.com/ManimCommunity/manim/pull/1088>`__: Added new svg files to documentation and imports
* `#1076 <https://github.com/ManimCommunity/manim/pull/1076>`__: Improve documentation for GraphScene
Testing
-------

* `#1019 <https://github.com/ManimCommunity/manim/pull/1019>`__: Refactored :meth:`~.Scene.play`
* `#1100 <https://github.com/ManimCommunity/manim/pull/1100>`__: Rewrote test cases to use sys.executable in the command instead of "python"
* `#1079 <https://github.com/ManimCommunity/manim/pull/1079>`__: Removed the hardcoded value, `manim`, in `test_version.py`
Infrastructure
--------------

* `#1071 <https://github.com/ManimCommunity/manim/pull/1071>`__: Enable pytest-cov based code coverage
Unlabeled
---------

* `#1164 <https://github.com/ManimCommunity/manim/pull/1164>`__: Fixed single PEP8 style in `cairo_renderer.py`
* `#1140 <https://github.com/ManimCommunity/manim/pull/1140>`__: Flake8 Compat & Code Cleanup
* `#1124 <https://github.com/ManimCommunity/manim/pull/1124>`__: Deprecated :class:`ShowCreation` in favor of :class:`Create`
* `#1110 <https://github.com/ManimCommunity/manim/pull/1110>`__: Deprecated SmallDot + OpenGLSmallDot
* `#1073 <https://github.com/ManimCommunity/manim/pull/1073>`__: Removed "one line summary" from PULL_REQUEST_TEMPLATE.md
```

</details>

**Comment by jsonvillanueva** [comment]
Fixed those few things I mentioned in my previous comment. Checkout the changes -- I'm sure you'll have more ideas on how to iimprove the tool/have better formatting/make it easier to use @naveen521kk

**Review by behackl — COMMENTED** [review]
Thank you very much for all of your efforts here, I like the generated output **a lot**!

I only have one remark: when stating it as "X people have contributed to this release", I feel that also reviewers should be included in that list -- after all, reviewing code is also a contribution.

We could also reword it slightly ("X people have authored changes for this release" maybe?), that would be more explicit in some sense, but if I could choose, I'd prefer including reviewers.

**Comment by kolibril13** [comment]
I also really like the level of automation we will reach with this pr, thanks a lot for your effort.
One thing to note, the output that you sent us in the gist looks very long:
image
where the previous changelog entries were quite short:
image
Just wanted to mention this. Only as an idea, maybe the content of the summary can be collapsed?
https://www.w3schools.com/howto/howto_js_collapsible.asp

**Comment by jsonvillanueva** [comment]
@behackl 
I agree, reviewers should be credited for their contributions. I *might* not have time to get to this.... I'll see what I can do, but I'll be away this weekend and would like to focus more on #1013 considering timing.

@kolibril13 
I really like that idea of collapsing the descriptions. I'm not sure atm how this works for rst, but I'll look into it as well.

**Comment by behackl** [comment]
> @behackl
> I agree, reviewers should be credited for their contributions. I _might_ not have time to get to this.... I'll see what I can do, but I'll be away this weekend and would like to focus more on #1013 considering timing.
> 
> This looks relevant though: https://pygithub.readthedocs.io/en/latest/github_objects/PullRequest.html?highlight=pullrequest#github.PullRequest.PullRequest.get_reviews
> 
> @kolibril13
> I really like that idea of collapsing the descriptions. I'm not sure atm how this works for rst, but I'll look into it as well.

If you'd rather work on #1013 that's perfectly fine. I have some time today and can investigate both of these issues.

**Comment by jsonvillanueva** [comment]
> 
> 
> > @behackl
> > I agree, reviewers should be credited for their contributions. I _might_ not have time to get to this.... I'll see what I can do, but I'll be away this weekend and would like to focus more on #1013 considering timing.
> > This looks relevant though: https://pygithub.readthedocs.io/en/latest/github_objects/PullRequest.html?highlight=pullrequest#github.PullRequest.PullRequest.get_reviews
> > @kolibril13
> > I really like that idea of collapsing the descriptions. I'm not sure atm how this works for rst, but I'll look into it as well.
> 
> If you'd rather work on #1013 that's perfectly fine. I have some time today and can investigate both of these issues.

By all means, go ahead! 

As for the actual change: The logic for where the authors are credited would need scratched/done differently -- I think it should be based on the :class:`PullRequest` using the above API call instead of using just the names from the commits -- which means it should be done after getting the list of PRs https://github.com/ManimCommunity/manim/pull/1138/files#diff-c461d965355498ee3bd152eda8723432a8d81b2bf09246f0cc8bf38c96d406a4R170

Would still need to investigate why #1075 is not in the list though

**Comment by behackl** [comment]
> Would still need to investigate why #1075 is not in the list though

... because the current logic searches for `"Merge pull request \\#(\\d*)"`, and #1075 has been merged with the custom message `Add OpenGL Renderer (#1075)`.

**Comment by behackl** [comment]
I've pushed a commit which makes the script also get the list of reviewers. It works locally, and I'm open for feedback.

If someone with a less experimental setup (I currently have to exclude support for Python 3.6 manually on my system, because otherwise a too low version of `numpy` would be installed) could merge the current master and run `poetry update` in order to resolve the lockfile conflict, that would be great.

I looked into collapsable sections for a bit (we could use something like https://pypi.org/project/sphinxcontrib-details-directive/, but would then have to deal with formatting I think; haven't tried it locally yet) but have to go and do some other things now. From my point of view, this PR could also be merged without support for collapsable sections.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (naveen521kk) [inline .github/PULL_REQUEST_TEMPLATE.md:16]: ask authors to nor remove the below comments.

File: scripts/dev_changelog.py
Comment (naveen521kk) [inline scripts/dev_changelog.py:24]:
```suggestion
    $ ./scripts/dev_changelog.py $GITHUB v0.3.0..v0.4.0 > 0.4.0-changelog.rst
```

File: scripts/dev_changelog.py
Comment (naveen521kk) [inline scripts/dev_changelog.py:41]: Is this even required?

File: scripts/dev_changelog.py
Comment (naveen521kk) [inline scripts/dev_changelog.py:47]:
```suggestion
    pat = r"^.*\t(.*)$"
```

File: scripts/dev_changelog.py
Comment (behackl) [reply scripts/dev_changelog.py:24]: Actually that should probably be something like
```suggestion
    $ ./scripts/dev_changelog.py $GITHUB v0.3.0..v0.4.0 -o 0.4.0-changelog.rst
```

File: scripts/dev_changelog.py
Comment (naveen521kk) [inline scripts/dev_changelog.py:73]:
```suggestion
    issues = re.findall(r".*\(\#(\d+)\)", merges)
```

File: scripts/dev_changelog.py
Comment (naveen521kk) [inline scripts/dev_changelog.py:80]:
```suggestion
    issues = re.findall(r"^.*\(\#(\d+)\)$", commits, re.M)
```

File: scripts/dev_changelog.py
Comment (jsonvillanueva) [reply scripts/dev_changelog.py:41]: Nope

File: scripts/dev_changelog.py
Comment (naveen521kk) [reply scripts/dev_changelog.py]: for now, it just prints it and redirecting it should be fine I think.

Comment (behackl) [reply scripts/dev_changelog.py]: When I used the script, it saved the changelog to `docs/source/changelog.rst`, it didn't just print it.

Comment (jsonvillanueva) [reply scripts/dev_changelog.py]: @behackl  is correct
I changed the logic to write to a file instead of printing so that we could use tqdm

Comment (naveen521kk) [reply scripts/dev_changelog.py]: Originally, it used to print now it is modified. I think the docstring must be changed

Comment (jsonvillanueva) [reply scripts/dev_changelog.py]: I just did @behackl suggestion and changed the usage --- everything else could have a docstring, but it's better to spend effort first implementing @kolibril13 suggestion before documenting. Unfortunately, I have to go now

File: docs/source/changelog.rst
Comment (naveen521kk) [inline docs/source/changelog.rst:1]: I think we should keep the old changelog somewhere

Comment (behackl) [reply docs/source/changelog.rst:1]: Indeed. We can maybe think about splitting the changelog into different files for different versions, but I am also against simply overriding the current file.

Comment (jsonvillanueva) [reply docs/source/changelog.rst:1]: This was my bad. I accidentally added these changes in 7d9d605. Feel free to revert this file for now.

I think changelog.rst should just be a `.. toctree::`  linking to the version of the generated changelog (e.g. 0.5.0-changelog.rst) and the older version should be renamed and/or split into multiple files? What do you guys think?

Comment (naveen521kk) [reply docs/source/changelog.rst:1]: I think that the script, writing it to a file, I think it should be better than we print it on screen and ask the release managers to place it accordingly.
What I think is to create a folder called changelog, have this as `changelog-0.1.0-0.4.0.rst` and newer changelogs will have `release-notes-0.5.0.rst`. Thoughts?

Comment (behackl) [reply docs/source/changelog.rst:1]: For the sake of consistency, I'd rather split all of the previous section into separate files. But otherwise I am all for it. 👍

Comment (jsonvillanueva) [reply docs/source/changelog.rst]: I reverted the changelog (95b02a4) in case we'd like to revisit it, but I've also gone ahead and done two small things ():

1. Split the previous section into separate files
2. Made the script output `{cur_release[1:]}-changelog.rst` by default (remove the `v` from the git tag/revision and place it in the appropriate directory under a `:reversed:` `:glob:`... might want to set a depth.
image

Comment (WampyCakes) [reply docs/source/changelog.rst]: > remove the v from the git tag/revision

What do you mean by this?

Comment (jsonvillanueva) [reply docs/source/changelog.rst]: https://github.com/ManimCommunity/manim/pull/1138/commits/be5baa70183c5acf86bddc497c029ee19d7227eb#diff-c461d965355498ee3bd152eda8723432a8d81b2bf09246f0cc8bf38c96d406a4R162

**Review by behackl — APPROVED** [review]
I think the current version of the script is good to go; it will certainly be better to use this script than generate the changelog manually for the next release.

If all open discussions are resolved and no further concerns are raised, I'll merge this in a few hours. (Or if someone else wants to do in the meantime: please feel free to do so.)

Further (substantial) improvements can be made in follow-up PRs.

**Comment by naveen521kk** [comment]
Thanks for the work here @jsonvillanueva and @behackl

**Pull request merged**
